### PR TITLE
[android] Ellipsize routing time & distance strings from the end

### PR DESCRIPTION
**Pull request by pastk**
Needed-for: #1528

### CODE REVIEW

**Comment by vng** [comment]
Can you provide a screenshot how a long time/distance string looks now?

**Comment by pastk** [comment]
> Can you provide a screenshot how a long time/distance string looks now?

Sure, please see it at https://github.com/organicmaps/organicmaps/issues/1528#issuecomment-996144098

**Review by vng — APPROVED** [review]

**Comment by biodranik** [comment]
Would it be better to reduce the font size and to fit everything, as we discussed? Or even better, to move the Start button down?

**Comment by pastk** [comment]
BTW what is the policy regarding min android version OM should support?
TextView font autosizing was added in Android 8.0 Oreo (API 26).

**Comment by pastk** [comment]
> Would it be better to reduce the font size and to fit everything, as we discussed? Or even better, to move the Start button down?

I don't see how to make a layout nicely looking yet compact by moving the Start down.. Do you mean move it down into the right bottom corner? Then the elevation change numbers will be above the button which is not very logical or good looking IMHO.

If you mean move the button down below the chart then it will increase the bottom pane height significantly..

I believe there was an idea discussed somewhere (can't find it in issues though) to make the elevation chart more detailed and useful. So maybe it'll be good to redesign it all together later. E.g. we can have an expandable bottom pane just like the PP is atm. In default height it'll display a small amount of basic info and when expanded it could show a better elevation profile, ETA, etc.

**Comment by biodranik** [comment]
> Do you mean move it down into the right bottom corner? Then the elevation change numbers will be above the button which is not very logical or good looking IMHO.

This. Is it easy to quickly test and preview it?

**Comment by biodranik** [comment]
> TextView font autosizing was added in Android 8.0 Oreo (API 26).

Can it be backported to 5.0+ using Compatibility classes?

**Comment by pastk** [comment]
> > Do you mean move it down into the right bottom corner? Then the elevation change numbers will be above the button which is not very logical or good looking IMHO.
> 
> This. Is it easy to quickly test and preview it?

No, not really (for me at least). The layout is split into several res files and I couldn't find out where relative position of the start button and routing numbers / alt chart are set.

**Comment by pastk** [comment]
> > TextView font autosizing was added in Android 8.0 Oreo (API 26).
> 
> Can it be backported to 5.0+ using Compatibility classes?

It should work for 4.0+ using AndroidX.

But unfortunately I couldn't make it work for the time/distance text at all. It just doesn't autosize. Maybe the font size is overriden somewhere or its due to the way layout stretching is set. I played with various settings, nothing worked. We might need someone more knowledgeable in Android layouts or wait till I learn :)

As another easy workaround (in addition to ellipsizing) I can suggest allowing the text to break into two lines when needed:
OM-route-distance-9km-bug-fixed-2lines

**Comment by biodranik** [comment]
> As another easy workaround (in addition to ellipsizing) I can suggest allowing the text to break into two lines when needed:

Looks good. Why ellipsing is necessary at all in this workaround?

**Comment by pastk** [comment]
I used non-breakable spaces (\u00A0) to control where the line breaks to avoid ugly output like
wo_nbsp

With non-breakable spaces:
w_nbsp

File: android/res/layout-land/routing_details.xml [hunk 1]
```diff
@@ -9,10 +9,10 @@
   <TextView
     android:id="@+id/time"
     android:singleLine="true"
-    android:ellipsize="middle"
+    android:ellipsize="end"
```
Comment (biodranik) [inline android/res/layout-land/routing_details.xml:12]: Should we remove it completely?

Comment (pastk) [reply android/res/layout-land/routing_details.xml:12]: Its needed for some cases still...

OM-route-distance-long-2lines-ellipsize

Comment (pastk) [reply android/res/layout-land/routing_details.xml:12]: BTW we can win some more space by removing whitespace between numbers and corresponding units,
e.g.
`11h 55min - 153km` - at the expense of slightly less readability

Comment (biodranik) [reply android/res/layout-land/routing_details.xml:12]: Let's try and see how it looks in normal and extreme cases.

Comment (pastk) [reply android/res/layout-land/routing_details.xml:12]: ### Large font:

nospace_1
nospace_2
nospace_3
nospace_4

### Normal font:
nospace_sm_1
nospace_sm_2

### It also affects display in navigation mode:
nospace_nav

I don't see any improvement TBH, so I suggest leave it as it was with whitespace before units (the current commit in the PR).

Comment (biodranik) [reply android/res/layout-land/routing_details.xml:12]: Ok, thanks. Let's make it ideal later )

**Review by biodranik — APPROVED** [review]

**Pull request merged**